tephra-api/src/main/java/co/cask/tephra/Transaction.java

```diff
@@ -108,18 +157,60 @@ public long getFirstShortInProgress() {
     return firstShortInProgress;
   }
 
+  /**
+   * Returns true if the given version corresponds to a transaction that was in-progress at the time this transaction
+   * started.
+   */
   public boolean isInProgress(long version) {
     return Arrays.binarySearch(inProgress, version) >= 0;
   }
 
+  /**
+   * Returns true if the given version is present in one of the arrays of excluded versions (in-progress and
+   * invalid transactions).
+   */
   public boolean isExcluded(long version) {
     return Arrays.binarySearch(inProgress, version) >= 0
       || Arrays.binarySearch(invalids, version) >= 0;
   }
 
+  /**
+   * Returns true if the the given version corresponds to one of the checkpoint versions in the current
+   * transaction.
+   */
+  public boolean isCheckpoint(long version) {
+    return Arrays.binarySearch(checkpointWritePointers, version) >= 0;
+  }
+
+  /**
+   * Returns whether or not the given version should be visible to the current transaction.  A version will be visible
+   * if it was successfully committed prior to the current transaction starting, or was written by the current
+   * transaction (using either the current write pointer or the write pointer from a prior checkpoint).
+   *
+   * @param version the data version to check for visibility
+   * @return true if the version is visible, false if it should be hidden (filtered)
+   *
+   * @see #isVisible(long, boolean) to exclude the current write pointer from visible versions.  This method always
+   *      includes the current write pointer.
+   */
   public boolean isVisible(long version) {
+    return isVisible(version, true);
+  }
+
+  /**
+   * Returns whether or not the given version should be visible to the current transaction.  A version will be visible
+   * if it was successfully committed prior to the current transaction starting, or was written by the current
+   * transaction (using either the current write pointer or the write pointer from a prior checkpoint).
+   *
+   * @param version the data version to check for visibility
+   * @param excludeCurrentWritePointer whether writes from the current write pointer should be visible
+   * @return true if the version is visible, false if it should be hidden (filtered)
+   */
+  public boolean isVisible(long version, boolean excludeCurrentWritePointer) {
```

Comment:
@ghelmling Regarding your other comment. You say I am proposing that 
{quote}
When a transaction is started it provides read-your-writes semantics, but as soon as a checkpoint is performed, writes at the current write pointer are no longer visible.
{quote}
I did not propose that. On the contrary I am proposing that the visibility within a transaction should not change over its lifetime. I am proposing that we introduce a visibility level in transactions, which is either include-own-writes or exclude-own-writes. That becomes a property of the transaction and does not change over its life time. When phoenix executes a sequence of statements in a transaction, it can start a exclude-own-writes transaction. After each statement, it checkpoints the transaction, thus making the writes of all previous statement visible to the subsequent statement. But for the entire life span, the transaction never reads it current write version. 

With that, the visibility is encapsulated inside the Transaction class, similar to how we encapsulate exclusion of invalid or in-progress transactions. If we should ever introduce a "dirty-read" mode, it would also be encapsulated in the Transaction through its isVisible() method. Or would you want to add another boolean flag named includeInProgressWrites? 

The advantage of relying solely on the Transaction for visibility is that existing datasets remain interoperable with Transaction Checkpoints: because the dataset (or tx-aware) itself does not need to implement any new logic for including or excluding the write version. 

With your approach, all existing tx-aware classes have to implement a new feature, namely, the option to exclude the write version. Otherwise they cannot be used in a check-pointed transaction. Encapsulation normally protect us from percolating changes into every related class. This is common practice in modern software engineering, and I am suggesting to keep it that way.